File: Raven.Client.Java/lib/src/test/java/raven/tests/mailinglist/Troy2.java
package raven.tests.mailinglist;

import static org.junit.Assert.assertEquals;

import org.junit.Test;

import raven.abstractions.basic.Reference;
import raven.abstractions.indexing.FieldIndexing;
import raven.client.IDocumentSession;
import raven.client.IDocumentStore;
import raven.client.RavenQueryStatistics;
import raven.client.RemoteClientTest;
import raven.client.document.DocumentStore;
import raven.client.indexes.AbstractIndexCreationTask;
import raven.linq.dsl.IndexExpression;
import raven.linq.dsl.expressions.AnonymousExpression;
import raven.tests.resultsTransformer.QStronglyTypedResultsTransformerTest_Order;

import com.mysema.query.annotations.QueryEntity;

public class Troy2 extends RemoteClientTest {

  @Test
  public void usingDefaultFieldWithSelectFieldsFails() throws Exception {
    try (IDocumentStore store = new DocumentStore(getDefaultUrl(), getDefaultDb()).initialize()) {
      new TesterSearch().execute(store);

      try (IDocumentSession s = store.openSession()) {
        Tester tester1 = new Tester(null, "FirstName 1", "LastName 1", "email1@test.com", "test1");
        Tester tester2 = new Tester(null, "FirstName 2", "LastName 2", "email2@test.com", "test2");

        QTroy2_Tester t = QTroy2_Tester.tester;

        s.store(tester1);
        s.store(tester2);
        s.saveChanges();

        Reference<RavenQueryStatistics> statsRef = new Reference<>();
        s.advanced().luceneQuery(Tester.class, TesterSearch.class)
            .waitForNonStaleResults()
            .statistics(statsRef)
            .usingDefaultField("Query")
            .openSubclause()
            .where("FirstName*")
            .closeSubclause()
            .andAlso()
            .whereEquals(t.email, "email1@test.com")
            .orderBy("+LastName")
            .skip(0)
            .take(10)
            .toList();

        assertEquals(1, statsRef.value.getTotalResults());

        s.advanced().luceneQuery(Tester.class, TesterSearch.class)
            .waitForNonStaleResults()
            .statistics(statsRef)
            .usingDefaultField("Query")
            .openSubclause()
            .where("FirstName*")
            .closeSubclause()
            .andAlso()
            .whereEquals(t.email, "email1@test.com")
            .orderBy("+LastName")
            .selectFields(PasswordOnly.class)
            .skip(0)
            .take(10)
            .toList();
        assertEquals(1, statsRef.value.getTotalResults());
      }
    }
  }

  public static class PasswordOnly {
    private String password;

    public String getPassword() {
      return password;
    }

    public void setPassword(String password) {
      this.password = password;
    }
  }

  public static class TesterSearch extends AbstractIndexCreationTask {

    @Override
    public String getIndexName() {
      return "Tester/Search";
    }

    @QueryEntity
    public static class SearchResult {
      private String query;
      private String fistName;
      private String lastName;
      private String email;
      private String password;
      public String getQuery() {
        return query;
      }
      public void setQuery(String query) {
        this.query = query;
      }
      public String getFistName() {
        return fistName;
      }
      public void setFistName(String fistName) {
        this.fistName = fistName;
      }
      public String getLastName() {
        return lastName;
      }
      public void setLastName(String lastName) {
        this.lastName = lastName;
      }
      public String getEmail() {
        return email;
      }
      public void setEmail(String email) {
        this.email = email;
      }
      public String getPassword() {
        return password;
      }
      public void setPassword(String password) {
        this.password = password;
      }

    }

    public TesterSearch() {

      QTroy2_TesterSearch_SearchResult s = QTroy2_TesterSearch_SearchResult.searchResult;
      QTroy2_Tester t = new QTroy2_Tester("testClass");

      map = IndexExpression.from(Tester.class)
          .select(
              new AnonymousExpression()
              .withTemplate(s.query, "new object[] { testClass.FirstName, testClass.LastName, testClass.Email }")
              .with(t.firstName)
              .with(t.lastName)
              .with(t.email)
              .with(t.password)
              );

      index(s.query, FieldIndexing.ANALYZED);
      index(s.fistName, FieldIndexing.DEFAULT);
      index(s.lastName, FieldIndexing.DEFAULT);
      index(s.email, FieldIndexing.DEFAULT);
      index(s.password, FieldIndexing.DEFAULT);
    }

  }


  @QueryEntity
  public static class Tester {


    public Tester() {
      super();
    }
    public Tester(String id, String firstName, String lastName, String email, String password) {
      super();
      this.id = id;
      this.firstName = firstName;
      this.lastName = lastName;
      this.email = email;
      this.password = password;
    }
    private String id;
    private String firstName;
    private String lastName;
    private String email;
    private String password;
    public String getId() {
      return id;
    }
    public void setId(String id) {
      this.id = id;
    }
    public String getFirstName() {
      return firstName;
    }
    public void setFirstName(String firstName) {
      this.firstName = firstName;
    }
    public String getLastName() {
      return lastName;
    }
    public void setLastName(String lastName) {
      this.lastName = lastName;
    }
    public String getEmail() {
      return email;
    }
    public void setEmail(String email) {
      this.email = email;
    }
    public String getPassword() {
      return password;
    }
    public void setPassword(String password) {
      this.password = password;
    }


  }
}

File: Raven.Tests.MailingList/Troy2.cs
using System.Collections.Generic;
using System.Linq;
using Raven.Abstractions.Indexing;
using Raven.Client;
using Raven.Client.Indexes;
using Raven.Tests.Common;

using Xunit;

namespace Raven.Tests.MailingList
{
	public class Troy2 : RavenTest
	{
		[Fact]
		public void UsingDefaultFieldWithSelectFieldsFails()
		{
			using (var store = NewDocumentStore())
			{
				new TesterSearch().Execute(store);
				using (var session = store.OpenSession())
				{

					var testClasses = new List<Tester>
					{
						new Tester
						{
							FirstName = "FirstName 1",
							LastName = "LastName 1",
							Email = "email1@test.com",
							Password = "test1"
						},
						new Tester
						{
							FirstName = "FirstName 2",
							LastName = "LastName 2",
							Email = "email2@test.com",
							Password = "test2"
						}
					};
					testClasses.ForEach(session.Store);
					session.SaveChanges();

					RavenQueryStatistics stats;
                    var query = session.Advanced.DocumentQuery<Tester, TesterSearch>()
									   .WaitForNonStaleResults()
									   .Statistics(out stats)
									   .UsingDefaultField("Query")
									   .OpenSubclause()
									   .Where("FirstName*")
									   .CloseSubclause()
									   .AndAlso()
									   .WhereEquals(x => x.Email, "email1@test.com")
									   .OrderBy("+LastName")
									   .Skip(0)
									   .Take(10)
									   .ToList();
					Assert.Equal(1, stats.TotalResults);

                    var selectFieldsQuery = session.Advanced.DocumentQuery<Tester, TesterSearch>()
								   .WaitForNonStaleResults()
								   .Statistics(out stats)
								   .UsingDefaultField("Query")
								   .OpenSubclause()
								   .Where("FirstName*")
								   .CloseSubclause()
								   .AndAlso()
								   .WhereEquals(x => x.Email, "email1@test.com")
								   .OrderBy("+LastName")
								   .SelectFields<PasswordOnly>()
								   .Skip(0)
								   .Take(10)
								   .ToList();
					Assert.Equal(1, stats.TotalResults);

				}
			}
		}

		public class TesterSearch : AbstractIndexCreationTask<Tester, TesterSearch.SearchResult>
		{

			public override string IndexName
			{
				get
				{
					return "Tester/Search";
				}
			}

			public class SearchResult
			{
				public string Query { get; set; }
				public string FirstName { get; set; }
				public string LastName { get; set; }
				public string Email { get; set; }
				public string Password { get; set; }
			}

			public TesterSearch()
			{
				Map = testClasses =>
						from testClass in testClasses
						select new
						{
							Query = new object[]
							{
								testClass.FirstName,
								testClass.LastName,
								testClass.Email
							},
							testClass.FirstName,
							testClass.LastName,
							testClass.Email,
							testClass.Password
						};

				Index(x => x.Query, FieldIndexing.Analyzed);
				Index(x => x.FirstName, FieldIndexing.Default);
				Index(x => x.LastName, FieldIndexing.Default);
				Index(x => x.Email, FieldIndexing.Default);
				Index(x => x.Password, FieldIndexing.Default);
			}
		}

		public class PasswordOnly
		{
			public string Password { get; set; }
		}

		public class Tester
		{
			public string Id { get; set; }
			public string FirstName { get; set; }
			public string LastName { get; set; }
			public string Email { get; set; }
			public string Password { get; set; }
		}
	}
}